File: c/diffusion.c
#include <stdio.h>
#include <stdlib.h>

#define mval(MEM,i,j,k) MEM[i*N*N+j*N+k]

int main (int argc, char** argv)
{
	const int maxsize = 10;
	const int N = maxsize;
	double* cube = malloc(N*N*N*sizeof(double));
	int i, j, k;
	// Partition Flag: (0) off, (1) on
	const int partitionflag = 1;

	// Zero the cube
	for (i = 0; i < maxsize; i++)
	{
		for (j = 0; j < maxsize; j++)
		{
			for (k = 0; k < maxsize; k++)
			{
				mval(cube,i,j,k) = 0.0;
			}
		}
	}

	double* partition = malloc(N*N*N*sizeof(double));
// If partition flag is on, construct the partition
	if (partitionflag == 1)
	{
		for(i = 0; i < maxsize; i++)
		{
			for (j = 0; j < maxsize; j++)
			{
				for (k = 0; k< maxsize; k++)
				{
					mval(partition,i,j,k) = 0.0;
				}
			}
		}
	}

	// construct the actual location of the partition
	if (partitionflag == 1)
	{
		int kp = maxsize / 2;
		for (i = 0; i < maxsize; i++)
		{
			for (j = maxsize - 1; j > (maxsize / 4); j--)
			{
				partition[i*N*N+j*N+k] = -1.0;
			}
		}
	}

	// Define variables
	double diffusion_coeff = 0.175;
	double room_dim = 5.0;
	double gas_speed = 250.0;
	double timestep = (room_dim / gas_speed) / (double) maxsize;
	double block_distance = room_dim / (double) maxsize;
	double dterm = diffusion_coeff * timestep / (block_distance*block_distance);

	// Initialize first cell
	cube[0] = 1.0e21;
	
	double time = 0.0;
	double ratio = 0.0;

	// Diffusion Process
	do
	{
		for (i = 0; i < maxsize; i++)
		{
			for (j = 0; j < maxsize; j++)
			{
				for (k = 0; k < maxsize; k++)
				{
					// Repeat fro each step of the cube face, if partition is not present then diffuse
					if (0 <= k-1 && k-1 < maxsize)
					{
						if (partitionflag == 1)
						{
							if ((partition[i*N*N+j*N+k] != -1.0) && (partition[i*N*N+j*N+(k-1)] != -1.0))
							{
								double change = (cube[i*N*N+j*N+k] - cube[i*N*N+j*N+(k-1)]) * dterm;
								cube[i*N*N+j*N+k] = cube[i*N*N+j*N+k] - change;
								cube[i*N*N+j*N+(k-1)] = cube[i*N*N+j*N+(k-1)] + change;
							}
						}
						else
						{
						double change = (cube[i*N*N+j*N+k] - cube[i*N*N+j*N+(k-1)]) * dterm;
						cube[i*N*N+j*N+k] = cube[i*N*N+j*N+k] - change;
						cube[i*N*N+j*N+(k-1)] = cube[i*N*N+j*N+(k-1)] + change;
						}
					}

					if (0 <= k+1 && k+1 < maxsize)
					{
						if (partitionflag == 1)
						{
							if ((partition[i*N*N+j*N+k] != -1.0) && (partition[i*N*N+j*N+(k+1)] != -1.0))
							{
								double change = (cube[i*N*N+j*N+k] - cube[i*N*N+j*N+(k+1)]) * dterm;
								cube[i*N*N+j*N+k] = cube[i*N*N+j*N+k] - change;
								cube[i*N*N+j*N+(k+1)] = cube[i*N*N+j*N+(k+1)] + change;
							}
						}
						else
						{
						double change = (cube[i*N*N+j*N+k] - cube[i*N*N+j*N+(k+1)]) * dterm;
						cube[i*N*N+j*N+k] = cube[i*N*N+j*N+k] - change;
						cube[i*N*N+j*N+(k+1)] = cube[i*N*N+j*N+(k+1)] + change;
						}
					}

					if (0 <= j-1 && j-1 < maxsize)
					{
						if (partitionflag == 1)
						{
							if ((partition[i*N*N+j*N+k] != -1.0) && (partition[i*N*N+(j-1)*N+k] != -1.0))
							{
								double change = (cube[i*N*N+j*N+k] - cube[i*N*N+(j-1)*N+k]) * dterm;
								cube[i*N*N+j*N+k] = cube[i*N*N+j*N+k] - change;
								cube[i*N*N+(j-1)*N+k] = cube[i*N*N+(j-1)*N+k] + change;
							}
						}
						else
						{
								double change = (cube[i*N*N+j*N+k] - cube[i*N*N+(j-1)*N+k]) * dterm;
								cube[i*N*N+j*N+k] = cube[i*N*N+j*N+k] - change;
								cube[i*N*N+(j-1)*N+k] = cube[i*N*N+(j-1)*N+k] + change;
						}
					}

					if (0 <= j+1 && j+1 < maxsize)
					{
						if (partitionflag == 1)
						{
							if ((partition[i*N*N+j*N+k] != -1.0) && (partition[i*N*N+(j+1)*N+k] != -1.0))
							{
								double change = (cube[i*N*N+j*N+k] - cube[i*N*N+(j+1)*N+k]) * dterm;
								cube[i*N*N+j*N+k] = cube[i*N*N+j*N+k] - change;
								cube[i*N*N+(j+1)*N+k] = cube[i*N*N+(j+1)*N+k] + change;
							}
						}
						else
						{
								double change = (cube[i*N*N+j*N+k] - cube[i*N*N+(j+1)*N+k]) * dterm;
								cube[i*N*N+j*N+k] = cube[i*N*N+j*N+k] - change;
								cube[i*N*N+(j+1)*N+k] = cube[i*N*N+(j+1)*N+k] + change;
						}
					}

					if (0 <= i-1 && i-1 < maxsize)
					{
						if (partitionflag == 1)
						{
							if ((partition[i*N*N+j*N+k] != -1.0) && (partition[(i-1)*N*N+j*N+k] != -1.0))
							{
								double change = (cube[i*N*N+j*N+k] - cube[(i-1)*N*N+j*N+k]) * dterm;
								cube[i*N*N+j*N+k] = cube[i*N*N+j*N+k] - change;
								cube[(i-1)*N*N+j*N+k] = cube[(i-1)*N*N+j*N+k] + change;
							}
						}
						else
						{
								double change = (cube[i*N*N+j*N+k] - cube[(i-1)*N*N+j*N+k]) * dterm;
								cube[i*N*N+j*N+k] = cube[i*N*N+j*N+k] - change;
								cube[(i-1)*N*N+j*N+k] = cube[(i-1)*N*N+j*N+k] + change;
						}
					}

					if (0 <= i+1 && i+1 < maxsize)
					{
						if (partitionflag == 1)
						{
							if ((partition[i*N*N+j*N+k] != -1.0) && (partition[(i+1)*N*N+j*N+k] != -1.0))
							{
								double change = (cube[i*N*N+j*N+k] - cube[(i+1)*N*N+j*N+k]) * dterm;
								cube[i*N*N+j*N+k] = cube[i*N*N+j*N+k] - change;
								cube[(i+1)*N*N+j*N+k] = cube[(i+1)*N*N+j*N+k] + change;
							}
						}
						else
						{
								double change = (cube[i*N*N+j*N+k] - cube[(i+1)*N*N+j*N+k]) * dterm;
								cube[i*N*N+j*N+k] = cube[i*N*N+j*N+k] - change;
								cube[(i+1)*N*N+j*N+k] = cube[(i+1)*N*N+j*N+k] + change;
						}
					}

				}
			}
		}

		// Update time
		time = time + timestep;

		double maxval = cube[0];
		double minval = cube[0];
		
		for (i=0; i < maxsize; i++)
		{
			for (j=0; j < maxsize; j++)
			{
				for (k=0; k < maxsize; k++)
				{
					// Determine maxval and minval 
					if (cube[i*N*N+j*N+k] > maxval)
						maxval = cube[i*N*N+j*N+k];
					// Partition values aren't changed so avoid zero
					if ((cube[i*N*N+j*N+k] < minval) && (cube[i*N*N+j*N+k] != 0.0))
						minval = cube[i*N*N+j*N+k];
				}
			}
		}

		// update ratio
		ratio = minval / maxval;
		printf("%f\n", ratio);		

	} while (ratio < 0.99); // termination condition

	printf("Box equilibrated in: %f\n", time);

	free(cube); // deallocate cube memory
	free(partition); //deallocate partition memory
}

File: c#/Diffusion.cs
using System;

// Checked for mass consistency by Dr. Pounds on 10/25/19

namespace Diffusion
{
	class Program
	{
	static void Main(string[] args)
	{
		// Variables for the max size of the cube and the cube itself
		int maxsize = 50;
		double[, ,] cube = new double [maxsize, maxsize, maxsize];

		// Partition Flag: (0) partition off, (1) partition on
		int partitionflag = 0; 

		// Zero the cube
		for (int i = 0; i < maxsize; i++)
		{
			for (int j = 0; j < maxsize; j++)
			{
				for (int k = 0; k < maxsize; k++)
				{
					cube[i, j, k] = 0.0;
				}
			}
		}

		double[, ,] partition = new double [maxsize, maxsize, maxsize];
		// If partition flag is on, construct the partition
		if (partitionflag == 1)
		{
			// Construct partition array
			for (int i = 0; i < maxsize; i++)
                	{
                        	for (int j = 0; j < maxsize; j++)
                        	{
                                	for (int k = 0; k < maxsize; k++)
                                	{
                                        	partition[i, j, k] = 0.0; // Zero partition array
                                	}
                        	}
                	}
		}

		if (partitionflag == 1)
		{
			int kp = maxsize / 2;
			for (int i = 0; i < maxsize; i++)
			{
				for (int j = maxsize - 1; j > (maxsize/4); j--)
				{
					partition[i, j, kp] = -1; // Spaces where the partition is present
				}
			}
		}

		// Variables
		double diffusion_coefficient = 0.175;
		double room_dimension = 5;								// Distance in meters
		double gas_speed = 250.0;								// Based on 100 g/mol at RT
		double timestep = (room_dimension / gas_speed) / maxsize;				// timestep in seconds
		double block_distance = room_dimension / maxsize;
		double dterm = diffusion_coefficient * timestep / (block_distance*block_distance);

		// Initialize first cell
		cube[0, 0, 0] = 1.0e21;

		//int pass = 0;
		double time = 0.0; 									// Tracks accumulated system time
		double ratio = 0.0;
		double change;

		do
		{
			for (int i=0; i<maxsize; i++)
			{
				for (int j=0; j<maxsize; j++)
				{
					for(int k=0; k<maxsize; k++)
					{
						// Repeat this for each cell, if not diffusing into partition diffuse
						if (0 <= k-1 && k-1 < maxsize)
						{
							if (partitionflag == 1)
							{
								if ((partition[i, j, k] != -1) && (partition[i, j, k-1] != -1))
								{
										change = (cube[i, j, k] - cube[i, j, k-1]) * dterm;
										cube[i, j, k] = cube[i, j, k] - change;
										cube[i, j, k-1] = cube[i, j, k-1] + change;
								}
							}
							else
							{
								change = (cube[i, j, k] - cube[i, j, k-1]) * dterm;
								cube[i, j, k] = cube[i, j, k] - change;
								cube[i, j, k-1] = cube[i, j, k-1] + change;
							}
						}

						if (0 <= k+1 && k+1 < maxsize)
						{
							if (partitionflag == 1)
							{
								if ((partition[i, j, k] != -1) && (partition[i, j, k+1] != -1))
								{
										change = (cube[i, j, k] - cube[i, j, k+1]) * dterm;
										cube[i, j, k] = cube[i, j, k] - change;
										cube[i, j, k+1] = cube[i, j, k+1] + change;
								}
							}
							else
							{
								change = (cube[i, j, k] - cube[i, j, k+1]) * dterm;
								cube[i, j, k] = cube[i, j, k] - change;
								cube[i, j, k+1] = cube[i, j, k+1] + change;
							}
						}

						if (0 <= j-1 && j-1 < maxsize)
						{
							if (partitionflag == 1)
							{
								if ((partition[i, j, k] != -1) && (partition[i, j-1, k] != -1))
								{
										change = (cube[i, j, k] - cube[i, j-1, k]) * dterm;
										cube[i, j, k] = cube[i, j, k] - change;
										cube[i, j-1, k] = cube[i, j-1, k] + change;
								}
							}
							else
							{
								change = (cube[i, j, k] - cube[i, j-1, k]) * dterm;
								cube[i, j, k] = cube[i, j, k] - change;
								cube[i, j-1, k] = cube[i, j-1, k] + change;
							}
						}

						if (0 <= j+1 && j+1 < maxsize)
						{
							if (partitionflag == 1)
							{
								if ((partition[i, j, k] != -1) && (partition[i, j+1, k] != -1))
								{
										change = (cube[i, j, k] - cube[i, j+1, k]) * dterm;
										cube[i, j, k] = cube[i, j, k] - change;
										cube[i, j+1, k] = cube[i, j+1, k] + change;
								}
							}
							else
							{
								change = (cube[i, j, k] - cube[i, j+1, k]) * dterm;
								cube[i, j, k] = cube[i, j, k] - change;
								cube[i, j+1, k] = cube[i, j+1, k] + change;
							}
						}

						if (0 <= i-1 && i-1 < maxsize)
						{
							if (partitionflag == 1)
							{
								if ((partition[i, j, k] != -1) && (partition[i-1, j, k] != -1))
								{
										change = (cube[i, j, k] - cube[i-1, j, k]) * dterm;
										cube[i, j, k] = cube[i, j, k] - change;
										cube[i-1, j, k] = cube[i-1, j, k] + change;
								}
							}
							else
							{
								change = (cube[i, j, k] - cube[i-1, j, k]) * dterm;
								cube[i, j, k] = cube[i, j, k] - change;
								cube[i-1, j, k] = cube[i-1, j, k] + change;
							}
						}

						if (0 <= i+1 && i+1 < maxsize)
						{
							if (partitionflag == 1)
							{
								if ((partition[i, j, k] != -1) && (partition[i+1, j, k] != -1))
								{
										change = (cube[i, j, k] - cube[i+1, j, k]) * dterm;
										cube[i, j, k] = cube[i, j, k] - change;
										cube[i+1, j, k] = cube[i+1, j, k] + change;
								}
							}
							else
							{
								change = (cube[i, j, k] - cube[i+1, j, k]) * dterm;
								cube[i, j, k] = cube[i, j, k] - change;
								cube[i+1, j, k] = cube[i+1, j, k] + change;
							}
						}
					}
				}
			}

			time = time + timestep; // update timestep

			// Check for mass consitency
			double sumval = 0.0;
			double maxval = cube[0, 0, 0];
			double minval = cube[0, 0, 0];
			for (int i=0; i<maxsize; i++)
			{
				for (int j=0; j<maxsize; j++)
				{
					for (int k=0; k<maxsize; k++)
					{
						maxval = Math.Max(cube[i, j, k], maxval);
						if (cube[i, j, k] != 0.0)	// Since the values where the partition is aren't changing, don't consider 0 for min
							minval = Math.Min(cube[i, j, k], minval);
						sumval += cube[i, j, k];
					}
				}
			}

			ratio = minval / maxval; // update ratio 
			Console.WriteLine($"{time} {ratio}");
		} while (ratio < 0.99);

		Console.WriteLine($"Box equilibrated in: {time} seconds of simulated time.");
	}
	}
}
